devito/equation.py

```diff
@@ -104,8 +123,10 @@ def _symbolic_functions(self):
 
     def xreplace(self, rules):
         """"""
-        return self.func(self.lhs.xreplace(rules), self.rhs.xreplace(rules),
-                         subdomain=self._subdomain)
+        eq = Eq(self.lhs.xreplace(rules), self.rhs.xreplace(rules),
+                subdomain=self._subdomain)
+        eq._substitutions = self._substitutions
```

Comment:
With the aforementioned `if` statement removed this line can also be removed since currently there's no need to hold any record of the desired substitutions after they've been carried out.

Reply:
>to hold any record of the desired substitutions after they've been carried out.

xreplace is not only called by the substitution. For example it is called in `inject` and that may contain FD expression that would loose the substitutions without keeping track of it.

Reply:
Yeah, I see. After `.stencil` is called (and hence we can be sure substitutions have been carried out) it would be nice if this property is expunged but I'll worry about that in the future when it becomes relevant.